homeassistant/components/lcn/cover.py

```diff
@@ -1,53 +1,60 @@
 """Support for LCN covers."""
+
 import pypck
 
-from homeassistant.components.cover import CoverEntity
-from homeassistant.const import CONF_ADDRESS
+from homeassistant.components.cover import DOMAIN as DOMAIN_COVER, CoverEntity
+from homeassistant.const import CONF_DOMAIN, CONF_ENTITIES
 
 from . import LcnEntity
-from .const import CONF_CONNECTIONS, CONF_MOTOR, CONF_REVERSE_TIME, DATA_LCN
-from .helpers import get_connection
+from .const import (
+    CONF_DOMAIN_DATA,
+    CONF_MOTOR,
+    CONF_REVERSE_TIME,
+    CONF_UNIQUE_DEVICE_ID,
+)
+from .helpers import get_device_connection
 
 PARALLEL_UPDATES = 0
 
 
-async def async_setup_platform(
-    hass, hass_config, async_add_entities, discovery_info=None
-):
-    """Setups the LCN cover platform."""
-    if discovery_info is None:
-        return
+def create_lcn_cover_entity(hass, entity_config, config_entry):
+    """Set up an entity for this domain."""
+    host_id = config_entry.entry_id
```

Comment:
Is this the same host_id that is also config_entry.title? (e.g. in cnServiceCall get_device_connection())

Where is the ConfigEntry.entry_id set? I can only see here https://github.com/home-assistant/core/blob/24ccdb55bb0a9e281e58acb9304763230db25d9f/homeassistant/config_entries.py#L147 where its set to a random number if there isn't a param, and i am unable to find where its set in this integration.  I see where the title and unique_id are set.

Assume entry_id and title are the same thing, would it make sense to align on entry_id as the way to match host_id? Also assuming entry_id is set to the same host_id, then it seems like that can simplify get_device_connection since you have all the arguments needed to lookup the config entry without extra translation steps inside.

Reply:
No, `config_entry.entry_id` id and `config_entry.title` are not the same. The `entry_id` is generated by HomeAssistant and assigned to the config_entry during a ConfigFlow. The `title` is first set during the ConfigFlow, but can also be changed in the UI integration panel.
Usually the `host_id` equals to the `entry_id` and the `host_name` equals to the `title`.

Reply:
How is the `host_id` equal to the `entry_id` if the `entry_id` is generated by home assistant?  Did the last sentence mean to say that the `host_id` is equalito the `unique_id`?

Perhaps we can look at the logic in `LcnServiceCall closer. I'm having trouble following since (1) `host_id` is compared to config_entry.title which is user editable which does not seem correct (instead preferring `config_entry.data[CONF_HOST]` and (2) `host_id` is equal to `entry_id` yet its also generated by HomeAssistant. 

I would have expected `LcnServiceCall` to check against either `unique_id` (since you assign it) or something in `config_entry.data`, but not `entry_id` (system assigned) and not `title` (user editable display).

Reply:
> How is the `host_id` equal to the `entry_id` if the `entry_id` is generated by home assistant? Did the last sentence mean to say that the `host_id` is equalito the `unique_id`?
> 
I should have used `host_name` here as it refers to the `config_entry.title`. I'll fix that to make the names consistent.

> I would have expected `LcnServiceCall` to check against either `unique_id` (since you assign it) or something in `config_entry.data`, but not `entry_id` (system assigned) and not `title` (user editable display).

The service calls are sent to the LCN hub hardware device but the first parameter is always an address of a LCN hardware device (connected to the hub via a wired bus system). The address is usually given as `hub.s000.m010` (s: segment, m: module).

I had to decide how to address the hub in a service call. Using `config_entry.entry_id` does not make any sense as it is not visible to the user (and inconvenient). The `config_entry.unique_id` is not visible to the user either.
So, only the `title` is left which corresponds to the host attribute in configuration.yaml. The benefit here is that we are not breaking anything here.

Reply:
I did some reconsideration on the way the LCN service calls are implemented . I think the implementation should be changed in the future but will lead to a breaking change (so I suggest to postpone it).

Explanation:
The architecture of LCN is a 1:n relationship from a hub to physically connected hardware devices. Then there is a 1:n relationship from each hardware device to the device’s hardware resources (sensors, relays, ...). The latter are represented by entities in Home Assistant.
All service calls are addressed to the hardware devices. Therefore it would be better not to implement them as integration services (using the address parameter). Maybe there is a possibility to create entities for the hardware devices and define [entity services](https://developers.home-assistant.io/docs/dev_101_services/#entity-services) them. 
I need to have a deeper look into it. This would definately solve the addressing issue because one could simply call the service on a registered hardware device entity.
That‘s why I marked this PR as draft to ensure that this will work...
@allenporter: I am interested in what you think about this?

Reply:
My impression was that CONF_HOST is also in the config entry, andi it seems like a more direct way to compare the host rather than relying on the title which is for display purposes.  Is async_step_import removing it? I'd suggest add it back and use that rather than title.

Reply:
Yes, async_step_import removes the host name and sets it as the configuration entry title. But to be honest, I wouldn‘t add it back to the config entry data. Once the configuration.yaml is removed the user would have to define the host name *and* the title during integration setup. He also has to remember the host name, because it wont show up anywhere (and also might be different from the title).
In the current state I just need an identifier of the integration that is visible to the user (only for the address parameter of the service calls). Maybe it’s comparable to the entitiy_ids. Once changed via the UI it will break automations, frontends, ...

Reply:
Are there any custom services in the integration that doesn't target entities? If not, it's easy to stick with entity_ids as user facing identifiers in service calls. The integration can keep track of what hardware is connected to what entities and the user just uses the correct entity_id.

If there are services that can't be represented and connected to a single entity, it's not as straight forward. Then it depends on what the service does what is the best way forward. Eg some services should maybe just be websocket commands instead of services, and exposed to the user in a config panel. Then we can show them as belonging to a Home Assistant device with a device id, that represent the hardware device.

If a service needs to be used in automations it can be exposed as a device action that uses the device registry api. This will also be exposed to the user as Home Assistant devices. The device registry requires a stable unique identifier per hardware device.

Reply:
I excluded the device registry stuff for the hub, the hardware devices and the resources (entities) to keep this PR small. But it will be part of one of the next PRs (as you can see in the animated GIF in the description).
All service calls target the hardware devices which are not represented as entities.
So a perfect solution would be to use the device actions you described in the last paragraph. I didn‘t know about that until now. I will take a closer look. 😀
But for the moment I would like to keep the service calls as they are until the future PR which includes the device registry. This ensures that there is no breaking change at this point.